### PR TITLE
 Tenants module

### PR DESCRIPTION
**Pull request by sebastienros**
Fully functional module, with database presets supported.
- Can change preset until the tenant is running
- Can change prefixes and hosts live
- Can disable, reenable, reload tenants

image

### CODE REVIEW

**Comment by sebastienros** [comment]
> If I create another tenant with the same subdirectory than an existing one, there is an error and the two tenants are messed up.

I don't understand this repro

**Comment by sebastienros** [comment]
> blank in the validation summary

This is a bug in MVC that is fixed in 1.1

**Comment by sebastienros** [comment]
@agriffard I have applied the changes. Please review.

File: src/Orchard.Cms.Web/Modules/Orchard.Tenants/Controllers/AdminController.cs [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+            }
+
+            if (!String.IsNullOrWhiteSpace(model.RequestUrlPrefix))
+            {
+                if (model.RequestUrlPrefix.Contains('/'))
+                {
+                    ModelState.AddModelError(nameof(EditTenantViewModel.RequestUrlPrefix), S["The url prefix can not contains more than one segment."]);
+                }
+
+                if (allShells.Any(x => String.Equals(x.Settings.RequestUrlPrefix.Trim(), model.RequestUrlPrefix.Trim(), StringComparison.OrdinalIgnoreCase)))
```
Comment (agriffard) [inline src/Orchard.Cms.Web/Modules/Orchard.Tenants/Controllers/AdminController.cs]: Settings.RequestUrlPrefix is null for default site.

Prevent null reference : Settings.RequestUrlPrefix?.Trim()

**Comment by sebastienros** [comment]
The bug that we saw during the meeting while reloading a disabled tenant is an issue in ASP.NET Core that disposes singleton objects too aggressively. This will be fixed once we move to 1.1.0.

File: src/Orchard.Cms.Web/Modules/Orchard.Tenants/Views/Admin/Index.cshtml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+<h1>@RenderTitleSegments(T["Tenants"])</h1>
+
+<nav class="admin-toolbar">
+    <div class="nav navbar-nav">
+        <a asp-route-action="Create" class="btn btn-primary" role="button">@T["Add Tenant"]</a>
+    </div>
+</nav>
+
+<form asp-action="Index"> @* the form is nessecary to generate and antiforgery token for the delete action *@
```
Comment (sebastienros) [inline src/Orchard.Cms.Web/Modules/Orchard.Tenants/Views/Admin/Index.cshtml]: It's a code comment ;)

File: src/Orchard.Cms.Web/Modules/Orchard.Tenants/Controllers/AdminController.cs [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+                return Unauthorized();
+            }
+
+            if (_currentShellSettings.Name != ShellHelper.DefaultShellName)
+            {
+                return Unauthorized();
+            }
+
+            var shellContext = GetShells()
+                .Where(x => String.Equals(x.Settings.Name, id, StringComparison.OrdinalIgnoreCase))
```
Comment (sebastienros) [inline src/Orchard.Cms.Web/Modules/Orchard.Tenants/Controllers/AdminController.cs]: It doesn't help a lot, makes things harder when I want to order them for display.

File: src/Orchard.Cms.Web/Modules/Orchard.Tenants/project.json [hunk 4]
```diff
@@ -0,0 +1,27 @@
+﻿{
+  "version": "2.0.0-*",
+  "dependencies": {
+    "Microsoft.Extensions.DependencyInjection": "1.0.0",
+    "Orchard.Environment.Navigation": "2.0.0-*",
+    "Orchard.Admin.Abstractions": "2.0.0-*",
+    "Orchard.Hosting": "2.0.0-*",
+    "Orchard.DisplayManagement": "2.0.0-*",
+    "Orchard.Recipes": "2.0.0-*"
```
Comment (sebastienros) [inline src/Orchard.Cms.Web/Modules/Orchard.Tenants/project.json]: DisplayManagement is necessary for the base razor page. 
I could remove Recipes.

File: src/Orchard.Cms.Web/Modules/Orchard.Users/Controllers/AccountController.cs [hunk 5]
```diff
@@ -125,7 +134,7 @@ private IActionResult RedirectToLocal(string returnUrl)
             }
             else
             {
-                return Redirect("~/");
+                return Redirect("~/" + _shellSettings.RequestUrlPrefix);
```
Comment (sebastienros) [inline src/Orchard.Cms.Web/Modules/Orchard.Users/Controllers/AccountController.cs]: I don't want to add unnecessary helper methods, even more that we'd still need to pass in the shellsettings. I still have hope we can make it automatic too using IHostingEnvironment.BasePath if we set it correctly and register a custom one in the shell service provider.

File: src/Orchard.Cms.Web/Modules/Orchard.Tenants/Controllers/AdminController.cs (hunk 3 above)
Comment (Jetski5822) [inline src/Orchard.Cms.Web/Modules/Orchard.Tenants/Controllers/AdminController.cs]: okay

**Comment by sebastienros** [comment]
Feel free to approve the review but don't merge it, I found an issue with DI when reloading tenants and I am currently fixing it,

File: src/Orchard.Cms.Web/Modules/Orchard.Tenants/Controllers/AdminController.cs [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+
+            if (!String.Equals(model.Name, ShellHelper.DefaultShellName, StringComparison.OrdinalIgnoreCase) && string.IsNullOrWhiteSpace(model.RequestUrlHost) && string.IsNullOrWhiteSpace(model.RequestUrlPrefix))
+            {
+                ModelState.AddModelError(nameof(EditTenantViewModel.RequestUrlPrefix), S["Host and Url Prefix can not be empty at the same time."]);
+            }
+        }
+
+        private IEnumerable<ShellContext> GetShells()
+        {
+            return _orchardHost.ListShellContexts().OrderBy(x => x.Settings.Name);
```
Comment (sebastienros) [inline src/Orchard.Cms.Web/Modules/Orchard.Tenants/Controllers/AdminController.cs]: Listing tenants in the admin is not a very hot path ... kind of doesn't matter in this case.

**Comment by sebastienros** [comment]
The branch is ready for prime time, fell free to merge.

**Comment by Jetski5822** [comment]
Approved.

**Comment by Jetski5822** [comment]
Merged

**Pull request merged**